Add working Node.js Native Messaging host

guest271314 opened the pull request:
Fixes https://github.com/mdn/content/issues/27129

<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

<!-- ✍️ Summarize your changes in one or two sentences -->

### Motivation

<!-- ❓ Why are you making these changes and how do they help readers? -->

### Additional details

<!-- 🔗 Link to release notes, vendor docs, bug trackers, source control, or other places providing more context -->

### Related issues and pull requests

<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request should be merged first, use "**Depends on:** #123" -->


<!-- 👷‍♀️ After submitting, go to the "Checks" tab of your PR for the build status -->

Commit:
Add working Node.js Native Messaging host

Fixes https://github.com/mdn/content/issues/27129

Commit:
Update index.md

Josh-Cena committed:
Merge branch 'main' into patch-2

Josh-Cena committed:
Apply suggestions from code review

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Josh-Cena commented on files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md:
```diff
@@ -235,67 +235,75 @@ Each message is serialized using JSON, UTF-8 encoded and is preceded with an uns
-
-      // Reset the read size and the queued chunks
-      flushChunksQueue();
+#!/usr/bin/env -S /full/path/to/node --max-old-space-size=6 --jitless --expose-gc --v8-pool-size=1
```
This makes the code not directly runnable. What's the consideration for not using `#!/usr/local/bin/node`? Note that if you put it in your package.json `scripts`, npm (and other package managers) is able to map this to the real Node path, even if Node does not live there.

guest271314 replied:
In general


- For people who fetch and use only the `node` executable. Contrary to some claims in the wild `node` executable works without any `npm`, `npx`, `node_modules` folder, or package.json file on the sytem;
- For the ability to pass options to the `node` executable on the shebang line

Specifically so users will have to actually write where the `node` executable is located, and don't become conplacent and expect `npm`, which is thrd-party software owned and controlled by GitHub, not Node.js organization, to just magically configure everything for them. Native Messaging _should be_ with hands on development.

guest271314 replied:
> This makes the code not directly runnable. What's the consideration for not using `#!/usr/local/bin/node`?

Depending on what I am doing, I might just get rid of everything except the `node` executable https://github.com/guest271314/download-node-nightly-executable.

Let's say I do something like this to get the nightly Node.js download. 
```
$ cd $HOME
$ mkdir bin
$ cd bin
$ deno
> let osArch = 'linux-x64';
let [node_nightly_build] = await (
  await fetch('https://nodejs.org/download/nightly/index.json')
).json();
let { version, files } = node_nightly_build;
let node_nightly_url = `https://nodejs.org/download/nightly/${version}/node-${version}-${osArch}.tar.gz`;
console.log(node_nightly_url);
# ...
$ wget --show-progress --progress=bar https://nodejs.org/download/nightly/v21.0.0-nightly2023092455fde47b1d/node-v21.0.0-nightly2023092455fde47b1d-linux-x64.tar.gz
# ...
cd $HOME
$ mousepad .profile
# ...
# node-v21.0.0-nightly2023092455fde47b1d-linux-x64
if [ -d "$HOME/bin/node-v21.0.0-nightly2023092455fde47b1d-linux-x64/bin" ] ; then
    PATH="$HOME/bin/node-v21.0.0-nightly2023092455fde47b1d-linux-x64/bin:$PATH"
fi
# ...
$ . ~/.profile
```

I never really _install_ Node.js onto the system. Don't have to.

Josh-Cena commented on files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md:
```diff
@@ -235,67 +235,75 @@ Each message is serialized using JSON, UTF-8 encoded and is preceded with an uns
+  output.set(header, 0);
+  output.set(json, 4);
+  process.stdout.write(output);
+  // Mitigate RSS increasing expotentially for multiple messages
+  // between client and host during same connectNative() connection
+  message = header = output = null;
+  global.gc();
```
I don't understand this part. Why is this necessary and why isn't the default GC reliable?

guest271314 replied:
`node` will contnually increase RSS on Linux when the options passed to `node` and overwriting variables to `null` and calling `gc()` does not happen. I'm pretty sure I left links in the code to issues where I try to solve that.

Josh-Cena commented on files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md:
```diff
@@ -235,67 +235,75 @@ Each message is serialized using JSON, UTF-8 encoded and is preceded with an uns
+if (process.stdout._handle) {
+  process.stdout._handle.setBlocking(true);
+}
+// https://github.com/denoland/deno/discussions/17236#discussioncomment-4566134
```
We need to explain what this function solves

guest271314 replied:
`node` is peculiar.  In some cases when `stdout` is greater than 65536 `node` won't write to `stdout`.

guest271314 replied:
@Josh-Cena For completeness, another implementation using `node` executable https://github.com/simov/native-messaging/blob/8e99d2a345ae94426a502d05aa5d57b966f6bc78/protocol.js.

guest271314 replied:
@Josh-Cena What's the hold up for merging now?

Commit:
Update index.md

Process stdin asynchronously

Commit:
Update files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Update index.md

Remove flags passed to node and process.stdout._handle.setBlocking(true); asynchronously process stdout using fsPromises.open(), filehandle.write().

Commit:
Update index.md

Remove comments above readFullAsync function.

Josh-Cena committed:
Update index.md

Josh-Cena reviewed (approved):
Sorry, I thought someone else would come around. This is a much more readable and modern example than the original one, so I'm going to merge it. Thank you for all the work!

Josh-Cena committed:
Use top-level await

Josh-Cena committed:
Update index.md

guest271314 reviewed (commented):
Any reason you removed the main function?

Josh-Cena commented:
@guest271314 This is already an ESM file. We can use top-level await.

guest271314 commented:
Yes, that is true. However, we might want to run other code inside of `main()` before we begin the `while` loop, and/or run other code after we call `main()`. Nothing is going to happen after that `while` loop without including the code in an asynchronous function. Just some functionality and extensibility to consider. We don't gain anything technically by removing the `main` function.

Josh-Cena commented:
I'm not sure what you mean—if you want to execute anything before the while, just put the code before the while. If you want to do something crazy complicated, just add the wrapper back. For the purpose of this example, without the async function is better.

guest271314 commented:
> For the purpose of this example, without the async function is better.

You havn't explained why you think without the `async` function is "better", which is a subjective term anyway, other than the fact top-level `await` is supported in Ecmascript Modules.

Was just curious why you chose to remove the function. Not something to create controversy over, anyway. Have a great day.

epragalakis commented on files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md:
```diff
@@ -235,67 +235,51 @@ Each message is serialized using JSON, UTF-8 encoded and is preceded with an uns
+  const stdout = await fs.open(`/proc/${process.pid}/fd/1`, "w");
+  await stdout.write(header);
+  await stdout.write(message);
+  await stdout.close();
```
hey @Josh-Cena 
long shot, but do you remember on why you had to do the above (lines 266-271)
instead of doing a `console.log` or `process.stdout.write` ?

Josh-Cena commented:
This was already there in @guest271314's version, and I don't know why. I would assume `process.stdout` is more portable than assuming the place of the FD, so if you want to change that I'm open to reviewing another PR.
